### PR TITLE
Remove Pyarrow dependency

### PR DESCRIPTION
**Pull request by safoinme**
## Describe changes
Remove Pyarrow while updating the NumPy materialize.

## Pre-requisites
Please ensure you have done the following:
- [X] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Review by schustmi — CHANGES_REQUESTED** [review]
The only problem I see here is backwards compatability. If we now (due to caching, or a user explicitly requesting) try load an artifact that was stored with an older version of ZenML, the loading fails. Can we maybe include some logic like this:

```python
if fileio.exists(new_file):
    # use the logic you now implemented
elif fileio.exists(old_files):
    try:
        import pyarrow
        # use old logic here
    except ImportError:
        # error message that tells them to install pyarrow to load legacy artifacts   
```

**Review by fa9r — APPROVED** [review]
The main change here is that we now use `.npy` format instead of `.parquet` when saving numpy arrays. Generally, this makes sense. 

However, if you have a big dataset, you might prefer to use `.parquet` over `.npy`, so it would be cool if we still had the old option available as well. This reminded me again of how awkward our materializer installation is right now: Since we don't want the added dependency, we would need to add the numpy parquet materializer as a separate integration, which also feels really weird.

@schustmi haven't we already discussed this over lunch the other day? Do you still remember what we came up with back then? 😅

Now, getting back on topic, the PR achieves what it should and the code looks good so 👍 from my side :)

File: src/zenml/materializers/numpy_materializer.py [hunk 1]
```diff
@@ -49,32 +52,54 @@ def handle_input(self, data_type: Type[Any]) -> "NDArray[Any]":
...
+                    input_stream = pa.input_stream(f)
+                    data = pq.read_table(input_stream)
+                vals = getattr(data.to_pandas(), DATA_VAR).values
+                return np.reshape(vals, shape_tuple)
+            except ImportError:
+                raise ImportError(
+                    "You have an old version of the NumpyMaterializer",
+                    "DataArtifact. stored in the artifact store.",
+                    "Please install pyarrow to read parquet files.",
+                    "You can do so by running `pip install pyarrow`.",
```
[the remark this answers is not in the document]
Comment (safoinme) [reply src/zenml/materializers/numpy_materializer.py:96]: All your suggestions are welcomed and important! thank you so much for always leaving reviews.

**Comment by safoinme** [comment]
@fa9r Thank you for the clarification on the new implementation and removal of `pyarrow`. I have a question, however, why is `parquet` considered better or preferable over `npy` or maybe `npz` which is the more compressed format of `npy` when it comes to big datasets?

**Comment by fa9r** [comment]
@safoinme You're right, `.npy` seems to be even more efficient than parquet for storing numpy arrays 👍

**Comment by fa9r** [comment]
In that case, ignore my comment from above, perhaps we will have this discussion in the future again if we ever choose to redesign the pandas materializer :D

**Comment by safoinme** [comment]
@fa9r I think that time has come hahaha! `pandas materialize` requires `pyarrow` which must be changed to `pickle`. while `parquet` is actually better performing in read and write when it's numerical data, pickle is better in categorical data pickle is actually better. according to this: [The Best Format to Save Pandas Data](https://towardsdatascience.com/the-best-format-to-save-pandas-data-414dca023e0d)

**Review by schustmi — CHANGES_REQUESTED** [review]
The numpy materializer looks great now!

I'm a little concerned by the pandas materializer (we don't usually use `pickle` as it can become problematic when loading the stored pickle file from a different OS/python version). Is there any other native way to store these instead of using pickle?

**Comment by safoinme** [comment]
@schustmi unfortunately alternative ways are either not native or not really performant in comparison to `pickle` or `parquet` however regarding the compatibility thing I don't think it's actually an os issue but more like a python issue which mainly occurs due to not specifying the protocol version which means newer python version (3.9>) will use the latest protocol `5` while the old one will use `4`. if we force this to 4  which is supported n all current python versions zenml support this may prevent compatibility issues.

**Review by fa9r — CHANGES_REQUESTED** [review]
@safoinme I'm not a big fan of pickle either because of the compatibility issues @schustmi mentioned. Perhaps we should have a short meeting today to discuss the pros/cons of the various formats pandas supports.

**Comment by fa9r** [comment]
Ah CI is already passing? Thought it would at least complain because `is_dataframe_or_series()` has no docstring. Nevermind, in that case looks all good from my side 🚀

File: src/zenml/materializers/pandas_materializer.py [hunk 2]
```diff
@@ -37,40 +42,77 @@ class PandasMaterializer(BaseMaterializer):
...
+        if fileio.exists(self.parquet_path):
+            if self.pyarrow_exists:
+                with fileio.open(self.parquet_path, mode="rb") as f:
+                    df = pd.read_parquet(f)
+            else:
+                raise ImportError(
+                    "You have an old version of a `PandasMaterializer` "
+                    "data artifact stored in the artifact store "
+                    "as a `.parquet` file, which requires `pyarrow` and "
+                    "`fastparquet`for reading, You can install `pyarrow` & "
```
[the remark this answers is not in the document]
Comment (safoinme) [reply src/zenml/materializers/pandas_materializer.py:95]: Apparently, it's optional `pyarrow` can be enough but `fastparquet` does exactly the same work as `pyarrow` there are big debates on StackOverflow about which one is better and produces fewer issues. so we will just stick with pyarrow for now I think

**Review by AlexejPenner — APPROVED** [review]
:seal:

**Pull request merged**